### PR TITLE
xray: fix the default sampling rate for AWS X-Ray tracer

### PR DESCRIPTION
**Pull request by suniltheta**
Signed-off-by: Sunil Narasimhamurthy sunnrs@amazon.com

Commit Message: fix the default sampling rate for AWS X-Ray tracer extension to 5%
Additional Description:
Risk Level: Low
Testing: unit test the LocalizedSamplingStrategy to check if the default sampling rate is 0.05 if no other localized rule is applied via configuration
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: docs/root/version_history/current.rst
Comment (tonya11en) [inline docs/root/version_history/current.rst:105]: Is there a reason this is under "bug fixes" and not "minor behavioral changes"?

Comment (suniltheta) [reply docs/root/version_history/current.rst:105]: As per the design proposal [#5644](https://github.com/envoyproxy/envoy/issues/5644) and documented behavior, the default sampling rate for AWS X-Ray tracer is expected to be 5%. It was a typo/bug in the code that the _DefaultRate_ value was 0.5 (50%) and not 0.05 (5%). Which will be addressed by this PR.

File: source/extensions/tracers/xray/localized_sampling.h
Comment (suniltheta) [reply source/extensions/tracers/xray/localized_sampling.h]: Thanks for the tip. I made this change.

**Review by tonya11en — DISMISSED** [review]

**Comment by suniltheta** [comment]
Hi @tonya11en due to the release v1.18 that went today there was a **merge conflict** in docs/root/version_history/current.rst file which I just resolved. Can you please review the changes again and approve? Thank you for your time.

File: docs/root/version_history/current.rst
Comment (tonya11en) [reply docs/root/version_history/current.rst:16]: Consider appending something like "as opposed to 50%" so it's clear that was the bug.

File: source/extensions/tracers/xray/localized_sampling.cc
Comment (tonya11en) [inline source/extensions/tracers/xray/localized_sampling.cc:13]: nit: End with a period.

You also can just start with `// Corresponds to...`.

It's not clear why only this variable gets an explanation, but the others do not. The JSON keys are obvious, but DefaultFixedTarget and SamplingFileVersion may as well get an explanation too.

File: source/extensions/tracers/xray/localized_sampling.h
Comment (tonya11en) [inline source/extensions/tracers/xray/localized_sampling.h:127]: Is this necessary? Why can't you just call `defaultRule().rate()`?

**Review by tonya11en — COMMENTED** [review]
Thanks for updating! This mostly LGTM, I just had questions/suggestions.

/wait-any

File: source/extensions/tracers/xray/localized_sampling.cc
Comment (suniltheta) [reply source/extensions/tracers/xray/localized_sampling.cc]: Added the comments (explanation) for the other two variables as well. 👍

And for the previous review comment `> Consider appending something like ...` I made the change in the latest commit. For some reason I was not able to directly reply to that review comment.

File: source/extensions/tracers/xray/localized_sampling.h
Comment (suniltheta) [reply source/extensions/tracers/xray/localized_sampling.h:127]: From earlier suggestion in this PR LocalizedSamplingStrategy::defaultManifest() function was made a const. 
`const LocalizedSamplingManifest& defaultManifest() const`.

So the return type is a **const** reference of LocalizedSamplingManifest object. To make a call `strategy.defaultManifest().defaultRule().rate()` in the UnitTest code, the function `LocalizedSamplingManifest::defaultRule()` was not expected to be called by a const reference of LocalizedSamplingManifest object.

Instead of making defaultRule function as const `LocalizedSamplingManifest::defaultRule const()` which in turn needed multiple changes where defaultRule function is called within the source code, I opted to crating a new function `double defaultRuleRate() const {}` which seemed like a less code change.

Comment (tonya11en) [reply source/extensions/tracers/xray/localized_sampling.h:127]: Can you not just make a copy of the default rule in the unit test from the cost ref?

Comment (suniltheta) [reply source/extensions/tracers/xray/localized_sampling.h]: > Is this necessary? Why can't you just call `defaultRule().rate()`?

It is not necessary. In the new commit, got rid of this function (`defaultRuleRate`) and instead making a copy of the default_manifest_ from the const ref

**Review by tonya11en — APPROVED** [review]
Thanks! It's left in a better state than it was found, too!

LGTM assuming CI passes and all that.

**Comment by suniltheta** [comment]
Once again thanks for the review and suggestions, especially the hints for making it more readable.

**Comment by suniltheta** [comment]
Hi @tonya11en, I resolved a new merge conflict in docs/root/version_history/current.rst file. Can you please approve once again?
Any tip to get attention from maintainers to get this PR merged, in a nice way?

File: docs/root/version_history/current.rst
Comment (tonya11en) [inline docs/root/version_history/current.rst:26]: rm extra line

**Review by tonya11en — CHANGES_REQUESTED** [review]
Looks like there's another conflict :(. There's a lot of churn right now in the release notes because we're setting up the new versions. Thanks for being patient.

I'll ping a maintainer once that's all fixed.

File: docs/root/version_history/current.rst
Comment (suniltheta) [reply docs/root/version_history/current.rst]: Removed this extra line and resolved the merge conflict.

**Review by tonya11en — APPROVED** [review]
@asraa since Matt's out of town, can another maintainer take the last pass?

**Comment by asraa** [comment]
Sure! Changes LGTM, would like to get an xray extension codeowner to check @abaptiste 
But if not ping me and I will approve.

**Comment by abaptiste** [comment]
@asraa @tonya11en LGTM

**Review by asraa — APPROVED** [review]

**Comment by suniltheta** [comment]
thanks @abaptiste @asraa @tonya11en for you time.

**Pull request merged**
